File: src/cmcandy/C_Language_Answers/_0028.cpp
#include <iostream>
#include <vector>

using namespace ::std;

//KMP 算法
class Solution
{
public:
    int strStr(string haystack, string needle)
    {
        if (needle.empty())
            return 0;
        int i = 0, m = haystack.size();
        int j = 0, n = needle.size();
        vector<int> nextVal = get_nextVal(needle);
        while (i<m&&j<n)
        {
            if(j==-1||needle[j]==haystack[i])i++,j++;
            else j = nextVal[j];
        }
        return j==n?i-j:-1;
    }

    vector<int> get_next(const string &pat)
    {
        int n=pat.size();
        vector<int> next(n+1,-1);
        int k=-1,j=0;
        while(j<n){
            if(k==-1||pat[k]==pat[j]) next[++j]==++k;
            else k = next[k];
        }
        return next;
    }

    vector<int> get_nextVal(const string &pat)
    {
        int n = pat.size();
        vector<int> nextVal(n, -1);
        int k = -1, j = 0;
        while (j < n - 1)
        {
            if (k == -1 || pat[k] == pat[j])
            {
                k++, j++;
                if (pat[j] != pat[k])
                    nextVal[j] = k;
                else
                    nextVal[j] = nextVal[k];
            }
            else
            {
                k = nextVal[k];
            }
        }
        return nextVal;
    }
};

File: src/cmcandy/Python_language_Answers/_0028.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        lhay = len(haystack)
        lnee = len(needle)
        if lnee == 0:
            return 0
        elif lhay < lnee:
            return -1

        # 感觉可以用动态规划？还是那个字符串匹配
        # KMP
        next = self.buildNext(needle)
        j = 0
        index = 0
        while index < lhay:
            if needle[j] == haystack[index]:
                j += 1
            else:
                # 如果没有匹配到
                while j > -1 and needle[j] != haystack[index]:
                    j = next[j]
                if j == -1:
                    j = 0
                elif needle[j] == haystack[index]:
                    j += 1
            index += 1
            if j == lnee:
                return index - lnee + 1
            
        return -1



    def buildNext(self,needle:str):
        length = len(needle)
        j = 0
        k = -1
        next = [ 0 for n in range(length)]
        next[0] = -1
        while j < length - 1 :
            if k == -1 or needle[j] == needle[k]:
                k += 1
                j += 1
                next[j] = k
            else:
                k = next[k]
        return next
    def buildNextVal(self,needle:str):
        length = len(needle)
        j = 0
        k = -1
        next = [ 0 for n in range(length)]
        next[0] = -1
        while j < length - 1 :
            if k == -1 or needle[j] == needle[k]:
                k += 1
                j += 1
                if (needle[j] != needle[k]): 
                    next[j] = k  
                else :
                    #因为不能出现p[j] = p[ next[j ]]，所以当出现时需要继续递归，k = next[k] = next[next[k]]  
                    next[j] = next[k]; 
                
            else:
                k = next[k]
        return next


if __name__ == '__main__':
    s = Solution()
    print(s.buildNext('aabaaf'))
    pass
